File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def maxPathSum(self, root: Optional[TreeNode]) -> int:
    def dfs(node):
      if not node: 
        return (0, -inf)
      (h1, p1), (h2, p2) = dfs(node.left), dfs(node.right)
      return max(h1, h2, 0) + node.val, max(p1, p2, max(h1,0) + max(h2,0) + node.val)
    return dfs(root)[1]
  # lambda self, node, res=(lambda x,y:y): (lambda h1,p1,h2,p2: res(max(h1,h2,0) + node.val, max(p1, p2, max(h1,0) + max(h2,0) + node.val)))(*self.maxPathSum(node.left, lambda *x:x), *self.maxPathSum(node.right, lambda *x:x)) if node else res(0, -inf)
      

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
  int DFS(TreeNode* root, int& ans) {
    if (!root) {
      return 0;
    }
    int l = max(DFS(root->left, ans), 0);
    int r = max(DFS(root->right, ans), 0);
    ans = max(ans, root->val + l + r);
    return root->val + max(l, r);
  }
  int maxPathSum(TreeNode* root) {
    int ans = -1000*3e4-1;
    DFS(root, ans);
    return ans;
  }
};